packages/storage-plus/src/deque.rs

```diff
@@ -0,0 +1,534 @@
+    }
+
+    #[inline]
+    fn set_tail(&self, storage: &mut dyn Storage, value: u32) {
+        self.set_meta_key(storage, TAIL_KEY, value);
+    }
+
+    /// Helper method for `tail` and `head` methods to handle reading the value from storage
+    fn read_meta_key(&self, storage: &dyn Storage, key: &[u8]) -> StdResult<u32> {
+        let full_key = namespaces_with_key(&[self.namespace], key);
+        storage
+            .get(&full_key)
+            .map(|vec| {
+                Ok(u32::from_be_bytes(
+                    vec.as_slice()
+                        .try_into()
+                        .map_err(|e| StdError::parse_err("u32", e))?,
+                ))
+            })
+            .unwrap_or(Ok(0))
+    }
+
+    /// Helper method for `set_tail` and `set_head` methods to write to storage
+    #[inline]
+    fn set_meta_key(&self, storage: &mut dyn Storage, key: &[u8], value: u32) {
+        let full_key = namespaces_with_key(&[self.namespace], key);
+        storage.set(&full_key, &value.to_be_bytes());
+    }
+
+    /// Tries to get the value at the given position
+    /// Used internally
+    fn get_at_unchecked(&self, storage: &dyn Storage, pos: u32) -> StdResult<Option<T>> {
+        let prefixed_key = namespaces_with_key(&[self.namespace], &pos.to_be_bytes());
+        may_deserialize(&storage.get(&prefixed_key))
+    }
+
+    /// Removes the value at the given position
+    /// Used internally
+    fn remove_at_unchecked(&self, storage: &mut dyn Storage, pos: u32) {
+        let prefixed_key = namespaces_with_key(&[self.namespace], &pos.to_be_bytes());
+        storage.remove(&prefixed_key);
+    }
+
+    /// Tries to set the value at the given position
+    /// Used internally when pushing
+    fn set_at_unchecked(&self, storage: &mut dyn Storage, pos: u32, value: &T) -> StdResult<()> {
+        let prefixed_key = namespaces_with_key(&[self.namespace], &pos.to_be_bytes());
+        storage.set(&prefixed_key, &to_vec(value)?);
+
+        Ok(())
+    }
+}
+
+#[cfg(feature = "iterator")]
+impl<'a, T: Serialize + DeserializeOwned> Deque<'a, T> {
+    pub fn iter(&self, storage: &'a dyn Storage) -> StdResult<DequeIter<T>> {
+        Ok(DequeIter {
+            deque: self,
+            storage,
+            start: self.head(storage)?,
+            end: self.tail(storage)?,
+        })
+    }
+}
+
+#[cfg(feature = "iterator")]
+pub struct DequeIter<'a, T>
+where
+    T: Serialize + DeserializeOwned,
+{
+    deque: &'a Deque<'a, T>,
+    storage: &'a dyn Storage,
+    start: u32,
+    end: u32,
+}
+
+#[cfg(feature = "iterator")]
+impl<'a, T> Iterator for DequeIter<'a, T>
+where
+    T: Serialize + DeserializeOwned,
+{
+    type Item = StdResult<T>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.start == self.end {
+            return None;
+        }
+
+        let item = self
+            .deque
+            .get_at_unchecked(self.storage, self.start)
+            .transpose()?;
```

Comment:
So if a piece of data that should be there is missing, we fail quietly by ending iteration? I wonder if we should produce an error instead. A queue with missing data in the middle sounds bad.

Reply:
Good catch, I will produce an error instead